### PR TITLE
SDK Support for literal field types

### LINKED ISSUE #19294: `count` function in SDK missing from types
### Describe the Bug

When trying to use `count()` on a relational field, TypeScript throws an error:

CleanShot 2023-07-29 at 12 19 13@2x

### To Reproduce

```ts
import { createDirectus, rest, readItems } from '@directus/sdk'

interface Example {
  id: string;
  related: number[] | Relational[];
}

interface Relational {
  id: number;
}

interface Schema {
  example: Example[];
  relational: Relational[];
}

const directus = createDirectus<Schema>('https://example.com').with(rest());

directus.request(readItems('example', {
  fields: ['count(related)']
});
```

```
Type '"count(related)"' is not assignable to type '"*" | keyof Example | QueryFieldsRelational<Schema, Example>'.ts(2322)

```

### Directus Version

v10.5.2

### Hosting Strategy

Self-Hosted (Docker Image)

### PR DESCRIPTION
**Pull request by br41nslug**
Fixes #19294

To allow for functions on specific field types i am introducing literal types support in the schema. Specifically for `json`,`csv` and `datetime` to generate types with functions included on only those specific types (instead of blindly suggesting all functions on any `string` type).

## Snippet for Testing

```ts
import { createDirectus, readItem, readSingleton, rest } from '@directus/sdk';

const client = createDirectus<MySchema>('http://localhost:8055/').with(rest());

const result1 = client.request(readItem('collection_a', '', {
    fields: ['count(m2m)', 'day(date_created)']
}))
const result2 = client.request(readItem('collection_b', '', {
    fields: ['count(value)', 'value']
}))
const result3 = client.request(readSingleton('collection_c', {
    fields: ['*']
}))

// Schema
export interface MySchema {
	collection_a: CollectionA[]; // regular collections are array types
	collection_b: CollectionB[];
	collection_c: CollectionC; // this is a singleton
	// junction collections are collections too
	collection_a_b_m2m: CollectionAB_Many[];
	collection_a_b_m2a: CollectionAB_Any[];
}

// collection A
export interface CollectionA {
	id: number;
	status: string;
	date_created: 'datetime' | null;
	// relations
	m2o: number | CollectionB;
	o2m: number[] | CollectionB[] | null;
	m2m: number[] | CollectionAB_Many[] | null;
	m2a: number[] | CollectionAB_Any[] | null;
}

// Many-to-Many junction table
export interface CollectionAB_Many {
	id: number;
	collection_a_id: number | CollectionA;
	collection_b_id: number | CollectionB;
}

// Many-to-Any junction table
export interface CollectionAB_Any {
	id: number;
	collection_a_id: number | CollectionA;
	collection: 'collection_b' | 'collection_c';
	item: string | CollectionB | CollectionC;
}

// collection B
export interface CollectionB {
	id: number;
	value: 'json' | null;
}

// singleton collection
export interface CollectionC {
	id: number;
	app_settings: string;
	something: 'csv';
}
```

### CODE REVIEW

File: sdk/src/types/functions.ts [hunk 1]
```diff
@@ -1,9 +1,18 @@
...
 /**
  * Available query functions
  */
+export type DateTimeFunctions = 'year' | 'month' | 'week' | 'day' | 'weekday' | 'hour' | 'minute' | 'second';
+export type ArrayFunctions = 'count';
+
 export type QueryFunctions = {
-	date: 'year' | 'month' | 'week' | 'day' | 'weekday' | 'hour' | 'minute' | 'second';
+	datetime: DateTimeFunctions;
 	array: 'count';
```
[the remark this answers is not in the document]
Comment (br41nslug) [reply sdk/src/types/functions.ts:13]: Ah nice catch on the copy-paste error 🚀

File: .changeset/funny-hornets-compete.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+"@directus/sdk": minor
+---
+
+Added Support for literal field types in the SDK
```
Comment (paescuj) [inline .changeset/funny-hornets-compete.md:5]:
```suggestion
Added support for literal field types in the SDK
```

File: sdk/src/types/output.ts [hunk 3]
```diff
@@ -77,3 +80,36 @@ export type ApplyNestedQueryFields<Schema extends object, Collection, Fields> =
...
+
+// Possible JSON types
+type JsonPrimitive = null | boolean | number | string;
+type JsonValue = JsonPrimitive | JsonPrimitive[] | { [key: string]: JsonValue };
+
+/**
+ * Output map for specific literal types
+ */
+export type FieldOutputMap = {
+	json: JsonValue | null;
```
Comment (paescuj) [inline sdk/src/types/output.ts:112]: `null` is already part of `JsonPrimitive`
```suggestion
	json: JsonValue;
```

**Review by paescuj — DISMISSED** [review]
In general the changes look good to me but it doesn't filter the return type of request anymore:

### Before



### After



---

Also if using a function, for example `day(date_created)`, the value won't be under the original field name but a suffixed one:
```json
{ "date_created_day": 26 }
```
This is not reflected in the types.

**Review by paescuj — APPROVED** [review]
Awesome work, @br41nslug ❤️ 

happy-overwhelmed

Works like a charm now:

**Pull request merged**